### ISSUE TITLE
Gamepad key mapping oddities in I.E. on Xbox One?

### ISSUE THREAD

**Issue by staxmanade** [winjs/winjs#1541]
Hey All,

I'm seeing what appears to be strange navigation behavior when playing with WinJS inside of a web page and browsing to it in the Xbox One version of I.E.

# Issues I'm seeing

1. Without calling the `navigator.getGamepads()` up/down/left/right nav works, but `A` selection doesn't work
2. When I call `navigator.getGamepads()` the keyCodes get all messed up (up/down/left/right don't work properly)
3. Not in this demo - but for some reason when I call `navigator.getGamepads()` and I apply the following key-code mapping hack, this fixes everything EXCEPT focus inside of ListView's doesn't appear to work properly. (The list view items are getting focused, but I can't see the focus style).

# Steps to reproduce:

1. Load this page up in the Xbox One Internet Explorer (in Plunker's embedded mode). 
2. Use the Controller to navigate around the focus panels below (notice how navigation works great)  
3. Select the "click me" panel below with the controller'a A button (notice nothing happens)
<br />
<br />
4. Now select the first button below "1. call navigator.getGamePads()"
5. Try to browse around the focus area - notice how the key mappings are wrong?
<br />
<br />
6. Now FORCE CLOSE IE and re-launch to get a fresh start on the page and select "2. Fix keyCodes"
7. Notice now how the navigation is fixed. AND the click me works? (you see a toastr "clickMeClicked" message)

Plunker repro page:

## Gamepad keyCodeMapp re-mapping hack:

```
((keys, keyCodeMap) => {
    var Keys = keys;

    keys.GamepadA = 193;
    keys.GamepadB = 194;
    keys.GamepadDPadDown = 202;
    keys.GamepadDPadLeft = 203;
    keys.GamepadDPadRight = 204;
    keys.GamepadDPadUp = 201;
    keys.GamepadLeftShoulder = 198;
    keys.GamepadLeftThumbstick = 207;
    keys.GamepadLeftThumbstickDown = 210;
    keys.GamepadLeftThumbstickLeft = 212;
    keys.GamepadLeftThumbstickRight = 211;
    keys.GamepadLeftThumbstickUp = 209;
    keys.GamepadLeftTrigger = 199;
    keys.GamepadMenu = 205;
    keys.GamepadRightShoulder = 197;
    keys.GamepadRightThumbstick = 208;
    keys.GamepadRightThumbstickDown = 214;
    keys.GamepadRightThumbstickLeft = 216;
    keys.GamepadRightThumbstickRight = 215;
    keys.GamepadRightThumbstickUp = 213;
    keys.GamepadRightTrigger = 200;
    keys.GamepadView = 206;
    keys.GamepadX = 195;
    keys.GamepadY = 196;

    // clear out the bad values.
    keyCodeMap.left.length = 0;
    keyCodeMap.right.length = 0;
    keyCodeMap.up.length = 0;
    keyCodeMap.down.length = 0;
    keyCodeMap.accept.length = 0;
    keyCodeMap.cancel.length = 0;

    // re-initialize
    // https://github.com/winjs/winjs/blob/master/src/js/WinJS/XYFocus.ts
    keyCodeMap.left.push(Keys.GamepadLeftThumbstickLeft, Keys.GamepadDPadLeft, Keys.NavigationLeft);
    keyCodeMap.right.push(Keys.GamepadLeftThumbstickRight, Keys.GamepadDPadRight, Keys.NavigationRight);
    keyCodeMap.up.push(Keys.GamepadLeftThumbstickUp, Keys.GamepadDPadUp, Keys.NavigationUp);
    keyCodeMap.down.push(Keys.GamepadLeftThumbstickDown, Keys.GamepadDPadDown, Keys.NavigationDown);
    keyCodeMap.accept.push(Keys.GamepadA, Keys.NavigationAccept);
    keyCodeMap.cancel.push(Keys.GamepadB, Keys.NavigationCancel);

})(WinJS.Utilities.Key, WinJS.UI.XYFocus.keyCodeMap);
```

Any Thoughts?

**Comment by toddpi314** [comment]
@staxmanade can you compare your measured gamepad codes against what WinJS.Utilities.Key.* detects?

On our apps, we do find that we have to manually register our keys for most platforms/browsers with something like this:

``` javascript
if (WinJS.UI.XYFocus.keyCodeMap.up.indexOf(WinJS.Utilities.Key.upArrow) == -1)
      WinJS.UI.XYFocus.keyCodeMap.up.push(WinJS.Utilities.Key.upArrow);
    if (WinJS.UI.XYFocus.keyCodeMap.down.indexOf(WinJS.Utilities.Key.downArrow) == -1)
      WinJS.UI.XYFocus.keyCodeMap.down.push(WinJS.Utilities.Key.downArrow);
    if (WinJS.UI.XYFocus.keyCodeMap.left.indexOf(WinJS.Utilities.Key.leftArrow) == -1)
      WinJS.UI.XYFocus.keyCodeMap.left.push(WinJS.Utilities.Key.leftArrow);
    if (WinJS.UI.XYFocus.keyCodeMap.right.indexOf(WinJS.Utilities.Key.rightArrow) == -1)
      WinJS.UI.XYFocus.keyCodeMap.right.push(WinJS.Utilities.Key.rightArrow);
```

**Comment by toddpi314** [comment]
If that doesn't work, your answer will most likely be found by console logging the details here:
https://github.com/winjs/winjs/blob/master/src/js/WinJS/XYFocus.ts#L661-L679

The actual keycodes expected are here: https://github.com/winjs/winjs/blob/master/src/js/WinJS/Utilities/_ElementUtilities.js

FYI, the getGamepads has been experimental cross browser for a few years now and you could be running into a native bug in Edge which relates to runtime hardware detect. (speculating here)

**Comment by staxmanade** [comment]
@toddpi314 

If I don’t call `navigator.getGamePads()` - the `WinJS.Utilities.Key.*` values are all mapped good (except I can’t get the A button to fire). But once I call `navigator.getGamePads()` once - IE seems to change all the key-code mappings. So I have to re-map the codes myself.

I did copy over the arrow key mapping into my plunker demo above. This was based on a port of the XYFocus example here: http://winjs.azurewebsites.net/#xyfocus which has the following mappings applied:

```
// Wire XYFocus up to the arrow keys
WinJS.UI.XYFocus.keyCodeMap.up.push(WinJS.Utilities.Key.upArrow);
WinJS.UI.XYFocus.keyCodeMap.down.push(WinJS.Utilities.Key.downArrow);
WinJS.UI.XYFocus.keyCodeMap.left.push(WinJS.Utilities.Key.leftArrow);
WinJS.UI.XYFocus.keyCodeMap.right.push(WinJS.Utilities.Key.rightArrow);
```

These are helpful and necessary in some of my testing. (But doesn't explain why calling `navigator.getGamepads()` re-writes the keycode mappings coming from I.E...

If I pretend that my "hack" is good (for my needs now) - I guess I need to figure out why the `ListView` is not displaying the focus state, even-though focus IS actually happening.

**Comment by toddpi314** [comment]
@staxmanade: you should also look into dynamically reading the codes off of the gamepad api:

```javascript
window.addEventListener("gamepadconnected", function(e) {
  var gp = navigator.getGamepads()[0];
  __gp.mapping__
});
```

There could be unique layouts Edge is using that is getting things all messed up.
https://developer.mozilla.org/en-US/docs/Web/API/Gamepad_API/Using_the_Gamepad_API

**Comment by jdalton** [comment]
The keycodes in the mapping are for things like keypress events and not the browsers gamepad API.

**Comment by staxmanade** [comment]
@jdalton I agree with what you're saying (for what they're supposed to do) - however they `keypress` codes get all messed up if you call the gamepad api (even once).

**Comment by jdalton** [comment]
Ah I missed that part. Thanks for clarifying.

**Comment by lakichemole** [comment]
Hi,
I'm developing apps for Xbox one...and I choose the use WinJS 4.4 (instead of 1.0) so far so good...still some fix todo but not that bad, we had the same GameController issue, for the listview, the invoked event is catched only if the user click on it or key.code == enter, so using A button won't work, an easy workaround on our side was to set the enter keycode to 193 (A button), the best way would have been to handle game pad button inside the listview.

Romain

**Comment by toddpi314** [comment]
@lakichemole FYI, you planning on Cert '16? If not, might want to contact GDNP to make sure you can ditch XboxJS to get through Cert.

**Comment by colouredFunk** [comment]
@staxmanade Did you ever solve why the ListView is not displaying the focus state, even-though focus IS actually happening.

**Comment by staxmanade** [comment]
I have been working on the app outside of xbox for quite some time - so have not taken any time to address issues that will come up on the xbox... may be some time before I do...
